src/Microsoft.VisualStudio.Threading/ThreadingTools.cs

```diff
@@ -51,6 +51,49 @@ public static bool ApplyChangeOptimistically<T>(ref T hotLocation, Func<T, T> ap
             return true;
         }
 
+        /// <summary>
+        /// Optimistically performs some value transformation based on some field and tries to apply it back to the field,
+        /// retrying as many times as necessary until no other thread is manipulating the same field.
+        /// </summary>
+        /// <remarks>
+        /// Use this overload when <paramref name="applyChange"/> requires a single item, as is common when updating immutable
+        /// collection types. By passing the item as a method operand, the caller may be able to avoid allocating a closure
+        /// object for every call.
+        /// </remarks>
+        /// <typeparam name="TData">The type of data.</typeparam>
+        /// <typeparam name="TItem">The type of item.</typeparam>
+        /// <param name="hotLocation">The field that may be manipulated by multiple threads.</param>
+        /// <param name="item">A data value to pass to <paramref name="applyChange"/>.</param>
+        /// <param name="applyChange">A function that receives both the unchanged value and <paramref name="item"/>, then returns the changed value.</param>
+        /// <returns>
+        /// <c>true</c> if the location's value is changed by applying the result of the <paramref name="applyChange"/> function;
+        /// <c>false</c> if the location's value remained the same because the last invocation of <paramref name="applyChange"/> returned the existing value.
+        /// </returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "0#")]
+        public static bool ApplyChangeOptimistically<TData, TItem>(ref TData hotLocation, in TItem item, Func<TData, TItem, TData> applyChange)
```

Comment:
I think the `item` parameter is typically called `state` in most other similar contexts. Can we call it `state` here, and rename `TItem` to `TArg`?